### PR TITLE
CFE-586 : Apply user defined tags on created azure resources

### PR DESCRIPTION
**Pull request by bharath-b-rh**
PR is for applying OCP default tag (`"kubernetes.io_cluster.{cluster_id}" = "owned"`) and user defined tags to all azure resources created.

### CODE REVIEW

File: pkg/tfvars/azure/azure.go [hunk 1]
```diff
@@ -135,6 +135,7 @@ func TFVars(sources TFVarsSources) ([]byte, error) {
 		VMNetworkingType:                masterConfig.AcceleratedNetworking,
 		RandomStringPrefix:              randomStringPrefixFunction(),
 		VMArchitecture:                  vmarch,
+		ExtraTags:                       masterConfig.Tags,
```
Comment (barbacbd) [inline pkg/tfvars/azure/azure.go:138]: I see that the machines have the tags passed to them, but is this `ExtraTags` field used yet? Some of the providers have the installconfig tags passed to the tfvars initiated in `assets/cluster/tfvars.go`, should they be added there too or did I miss those tags?

**Review by jhixson74 — CHANGES_REQUESTED** [review]
Looking through the terraform changes, I don't see a need for local variables. I would pass the variables as passed into the top level variables file into each module. You also need to format the terraform using hack/tf-fmt.sh so that we don't get all these whitespace adjustments. Please squash the commits into a single commit.

File: pkg/tfvars/azure/azure.go (hunk 1 above)
Comment (bharath-b-rh) [reply pkg/tfvars/azure/azure.go:138]: Thank you for reviewing @barbacbd!

> I see that the machines have the tags passed to them, but is this ExtraTags field used yet?

Not until now, but with this change ExtraTags will be used for adding user-defined tags to azure resources created by installer.

ExtraTags is part of the azure platform specific config which is made available in the platform specific terraform variable file with the name `azure_extra_tags`. `Generate()` method in `assets/cluster/tfvars.go` generates the platform terraform variable file by calling `azure.TFVars()` method. azure_extra_tags is used in terraform code for applying tags it's initialized with to the azure resources created.

**Comment by bharath-b-rh** [comment]
Thank you for reviewing @jhixson74!

> Looking through the terraform changes, I don't see a need for local variables. I would pass the variables as passed into the top level variables file into each module.
```
locals {
  tags = merge(
    {
      "kubernetes.io_cluster.${var.cluster_id}" = "owned"
    },
    var.azure_extra_tags,
  )
  description = "Created By OpenShift Installer"
  # At this time min_tls_version is only supported in the Public Cloud and US Government Cloud.
  environments_with_min_tls_version = ["public", "usgovernment"]
}
```
Existing code defines a local variable containing both the ocp and user-defined tags. Shall I make changes in the terraform platform variable file generation code [here](https://github.com/openshift/installer/blob/bcba5f4374374351c2a9484cb764c21b3f91bb2f/pkg/tfvars/azure/azure.go#L138) to do the merge of both the tag list and to use it as is in terraform modules?

**Comment by bharath-b-rh** [comment]
> You also need to format the terraform using hack/tf-fmt.sh so that we don't get all these whitespace adjustments.
`ci/prow/tf-fmt` workflow was failing due to formatting and I used `hack/tf-fmt.sh` to get the diff and applied the changes suggested by the tf-fmt.

**Comment by bharath-b-rh** [comment]
> Please squash the commits into a single commit.

Could I do it after all comments are resolved, so that I can squash any new commits which maybe required?

**Comment by barbacbd** [comment]
> Thank you for reviewing @jhixson74!
> 
> > Looking through the terraform changes, I don't see a need for local variables. I would pass the variables as passed into the top level variables file into each module.
> 
> ```
> locals {
>   tags = merge(
>     {
>       "kubernetes.io_cluster.${var.cluster_id}" = "owned"
>     },
>     var.azure_extra_tags,
>   )
>   description = "Created By OpenShift Installer"
>   # At this time min_tls_version is only supported in the Public Cloud and US Government Cloud.
>   environments_with_min_tls_version = ["public", "usgovernment"]
> }
> ```
> 
> Existing code defines a local variable containing both the ocp and user-defined tags. Shall I make changes in the terraform platform variable file generation code [here](https://github.com/openshift/installer/blob/bcba5f4374374351c2a9484cb764c21b3f91bb2f/pkg/tfvars/azure/azure.go#L138) to do the merge of both the tag list and to use it as is in terraform modules?

@bharath-b-rh I would do it there yes. [AWS](https://github.com/openshift/installer/blob/master/pkg/tfvars/aws/aws.go#L80) actually does this explicitly (but probably doesn't currently need to since its just a complete match).

**Comment by barbacbd** [comment]
> > PR is for applying OCP default tag ("kubernetes.io_cluster.{cluster_id}" = "owned") and user defined tags to all azure resources created.
> 
> Can you point me to where in this PR we are setting the cluster ID tag? This is generally applied as one of the OpenShift reserved tags right? Where do we populate the infrastructure object with the user tags? We need to make sure we are excluding any openshift related tags there else they will count for the users tag count, I couldn't see that logic anywhere either

@JoelSpeed it looks like it was combined with other tags locally [here](https://github.com/openshift/installer/pull/6354/files#diff-b24d917166f09f74bbbd79611ba68631c35532fb81487580b8b2d54a4ccdc413R7). With John's suggested changes it would probably be moved to `tfvars/azure/azure.go`.

**Comment by bharath-b-rh** [comment]
> > PR is for applying OCP default tag ("kubernetes.io_cluster.{cluster_id}" = "owned") and user defined tags to all azure resources created.
> 
> Can you point me to where in this PR we are setting the cluster ID tag? This is generally applied as one of the OpenShift reserved tags right? Where do we populate the infrastructure object with the user tags? We need to make sure we are excluding any openshift related tags there else they will count for the users tag count, I couldn't see that logic anywhere either

Code for adding openshift reserved tag is in the terraform code [here](https://github.com/openshift/installer/pull/6354/files#diff-b24d917166f09f74bbbd79611ba68631c35532fb81487580b8b2d54a4ccdc413R5-R10), which already existed for adding tags to Azure ResourceGroup, hence not showing in the diff. 
openshift reserved tags are not included in the resourceTags of Infrastructure object, each operator should add openshift reserved tags to the retrieved list before applying on the resource. 
openshift/installer#6310 has the changes for populating Infrastructure resource tags.

**Comment by bharath-b-rh** [comment]
> @bharath-b-rh I would do it there yes. [AWS](https://github.com/openshift/installer/blob/master/pkg/tfvars/aws/aws.go#L80) actually does this explicitly (but probably doesn't currently need to since its just a complete match).

comment incoporated.

**Comment by JoelSpeed** [comment]
> Code for adding openshift reserved tag is in the terraform code [here](https://github.com/openshift/installer/blob/master/data/data/azure/bootstrap/main.tf#L5-L10), which already existed for adding tags to Azure ResourceGroup, hence not showing in the diff.
openshift reserved tags are not included in the resourceTags of Infrastructure object, each operator should add openshift reserved tags to the retrieved list before applying on the resource.
https://github.com/openshift/installer/pull/6310 has the changes for populating Infrastructure resource tags.
>
> On incorporating changes suggested by @jhixson74, as mentioned by @barbacbd merging openshift and user tags is now handled [here](https://github.com/openshift/installer/pull/6354/files#diff-2674c05dd65332988a049ec9d2e0fbd7c0ab5837902c0695c6f507fc3629c1d9R111-R117)

Ok yep, so where do we populate the infrastructure resource tags in this PR? Otherwise what you're saying here I agree with

**Comment by bharath-b-rh** [comment]
> Ok yep, so where do we populate the infrastructure resource tags in this PR? Otherwise what you're saying here I agree with

This PR doesn't have the changes for generating Infrastructure CR with resourceTags, this has changes for adding tags to the resources created by installer.

[openshift/installer#6310](https://github.com/openshift/installer/pull/6310) has the changes for populating Infrastructure resource tags.

**Comment by barbacbd** [comment]
I think this looks pretty good to me, but wait to tag this for @jhixson74

**Review by barbacbd — COMMENTED** [review]
/lgtm

**Pull request merged**